rasterio python quickstart
>reading and writing data files using rasterio
>
>use GeoTIFF format
>
>file is a worlclim imagery of kitui county in Geographic crs, i.e. WGS84

In [1]:
pip install rasterio

Note: you may need to restart the kernel to use updated packages.


# Opening a dataset in reading mode

In [13]:
import rasterio
#open the file
#open() function takes a path string and returns an opened dataset
dataset=rasterio.open('data/reproj_temp_kitui.tif')

Dataset objects have some of the same attributes as Python file objects.

In [14]:
#check the name
dataset.name

'data/reproj_temp_kitui.tif'

In [15]:
#check the mode
dataset.mode

'r'

# Dataset Attributes
access prpoerties of raster data through attributes of opened dataset object

In [16]:
#check the band count
dataset.count

1

dataset objects have bands
>dataset band - an array of values that represent the patrial distribution of a single variable in 2D space
>
>all band arrays of a dataset have the same no of rows and columns.

In [17]:
#check the array width
dataset.width

8

In [18]:
#check array height
dataset.height

17

# Dataset Georeferencing
GIS raster datasets are mapped to regions on the earth's surface

our example covers the world from 37.667 to 39.0 degrees, left to right, and 3.0 to 0.1667 degrees bottom to top

In [19]:
#check the spatial bounding box for our dataset
dataset.bounds

BoundingBox(left=351523.5507, bottom=9668321.852, right=499079.0256, top=9981877.2361)

Geospatial transform
>a fundamental attribute 
> gives the value of bounds

In [20]:
dataset.transform

Affine(18444.434362499996, 0.0, 351523.5507,
       0.0, -18444.434358823495, 9981877.2361)

Dataset Transform
>an affine transformation matrix
>
>maps pixel locations in (row/col) coords to (x/y) spatial positions

In [21]:
#spatial position of upper left corner=matrix * (0,0)
dataset.transform *(0,0)

(351523.5507, 9981877.2361)

In [22]:
#spatial position of lower right corner = affine matrix * (width, height)
dataset.transform * (dataset.width, dataset.height)

(499079.0256, 9668321.852)

Dataset CRS
>gives the coordinates i.e the numbers 351523.5507 meters

In [23]:
dataset.crs

CRS.from_epsg(21037)

'EPSG 21037' 
>identifies a particular crs: UTM Zone 37S, which covers a part of the southern hemisphere, including parts of Africa and the Indian Ocean
>
>using the crs attribute and transform, the georeferencing of a raster dataset is described and a dataset can be compared to othe GIS datasets

# Reading raster data
using the band's index number, we can access data from a raster band

bands are indexed from 1, following the GDAL convention.

In [29]:
#check how many bands are in the dataset
dataset.indexes


(1,)

In [30]:
band1 = dataset.read(1)
band1

array([[-3.4000000e+38, -3.4000000e+38, -3.4000000e+38,  1.7971500e+01,
         1.8388750e+01, -3.4000000e+38, -3.4000000e+38, -3.4000000e+38],
       [-3.4000000e+38, -3.4000000e+38,  1.7146500e+01,  1.6984751e+01,
         1.7568501e+01,  1.8632750e+01, -3.4000000e+38, -3.4000000e+38],
       [-3.4000000e+38,  1.6778500e+01,  1.5796500e+01,  1.6490000e+01,
         1.7558250e+01,  1.8552500e+01, -3.4000000e+38, -3.4000000e+38],
       [-3.4000000e+38,  1.6758249e+01,  1.5932500e+01,  1.7106750e+01,
         1.7831249e+01,  1.8738251e+01,  1.9692751e+01, -3.4000000e+38],
       [-3.4000000e+38,  1.5766000e+01,  1.6094000e+01,  1.7356750e+01,
         1.8530001e+01,  1.9297251e+01,  2.0114250e+01, -3.4000000e+38],
       [-3.4000000e+38,  1.5123750e+01,  1.5751750e+01,  1.7250250e+01,
         1.8812250e+01,  1.9788500e+01,  2.0401251e+01,  2.1014750e+01],
       [ 1.5371250e+01,  1.5319250e+01,  1.6208000e+01,  1.8117250e+01,
         1.9103251e+01,  1.9799999e+01,  2.0625000e+01,  2

access array values by their row, column index

In [31]:
band1[dataset.height// 2, dataset.width // 2]

np.float32(19.731)